### PR TITLE
Implement first version of PSI parser

### PR DESCRIPTION
**Pull request by e2e4b6b7**

### CODE REVIEW

File: plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java [hunk 1]
```diff
@@ -7,10 +7,10 @@
 import java.io.IOException;
```
Comment (e2e4b6b7) [inline plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java:7]: No logical changes in file (only for testing)

File: plugin/src/test/java/org/refactoringminer/test/TestAllRefactorings.java [hunk 2]
```diff
@@ -1,16 +1,20 @@
 package org.refactoringminer.test;
```
Comment (e2e4b6b7) [inline plugin/src/test/java/org/refactoringminer/test/TestAllRefactorings.java:1]: No logical changes in file (only for testing)

File: plugin/src/test/java/org/refactoringminer/test/TestBuilder.java [hunk 3]
```diff
@@ -37,23 +37,25 @@
     private Counter c;// = new Counter();
```
Comment (e2e4b6b7) [inline plugin/src/test/java/org/refactoringminer/test/TestBuilder.java:39]: No logical changes in file (only for testing)

**Comment by e2e4b6b7** [comment]
The most important files is UMLModelASTReader and Visitor. Also important UMLType with TypeUtils and VariableDeclaration. Other seems simple.

File: plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java [hunk 4]
```diff
@@ -78,7 +74,7 @@ private static void prepareFSERefactorings(TestBuilder test, BigInteger flag)
     private static boolean isAdded(Refactoring refactoring, BigInteger flag) {
         try {
             BigInteger value = Enum.valueOf(Refactorings.class, refactoring.type.replace(" ", "")).getValue();
-            return value.and(flag).compareTo(BigInteger.ZERO) == 1;
+            return value.and(flag).compareTo(BigInteger.ZERO) > 0;
```
Comment (onewhl) [inline plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java:77]: what is the reason to change it? o_o

File: refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java [hunk 5]
```diff
@@ -1,41 +1,65 @@
...
+            this.startColumn = 0;
+            this.endColumn = 0;
         }
-        this.endColumn = cu.getColumnNumber(endOffset);
-        //convert to 1-based
-        if (this.endColumn > 0) {
-            this.endColumn += 1;
+    }
+
+    public static TextRange getEclipseRange(@NotNull PsiElement node) {
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java:63]: do we need it only for comparing with the original tool?

File: refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java [hunk 6]
```diff
@@ -36,707 +43,223 @@
...
-                List<OperationInvocation> list = new ArrayList<>();
-                list.add(invocation);
-                anonymousMethodInvocationMap.put(nodeAsString, list);
+    @Override
+    public void visitElement(@NotNull PsiElement element) {
+        boolean goInSubtree = true;
+        if (element instanceof PsiArrayAccessExpression) {
+            String source = Formatter.format(element);
+            arrayAccesses.add(source);
+            if (!stackAnonymous.isEmpty()) {
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java]: why do we check stackAnonymous here and in the checkings below?

File: plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java (hunk 4 above)
Comment (e2e4b6b7) [reply plugin/src/test/java/org/refactoringminer/test/RefactoringPopulator.java:77]: Idea warning. It is documented for BigInteger that it is 1 but as I know it is not guaranteed in other cases so it is common practice to compare with 0

File: refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java (hunk 5 above)
Comment (e2e4b6b7) [reply refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java:63]: Hmm. You say that ranges matters in some metrics and I don't know in what cases it's important. If it's not important most of places, I'll concentrate on other differences. I planned to do as eclipse do in the and when I'll finish everything, remove cases which not affect any metrics

File: refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java (hunk 6 above)
Comment (e2e4b6b7) [reply refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java]: Logic or style? 
Logic is that we want to remove statements of Anonymous Classes in some places and we want to check if we in it. 
Style is that extracting it to the variable not makes code shorter or simpler. And some cases proceeds in extracted function so they need to check that condition again

File: refactoringminer/src/main/java/gr/uom/java/xmi/TypeUtils.java [hunk 7]
```diff
@@ -19,10 +21,7 @@ public static UMLType extractType(PsiFile file, String filePath, PsiNewExpressio
             PsiElement[] children = newExpression.getChildren();
             if (children[3] instanceof PsiKeyword) {
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/TypeUtils.java]: is it always third children?

File: refactoringminer/src/main/java/gr/uom/java/xmi/FormattingVisitor.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import org.jetbrains.annotations.NotNull;
+import java.util.regex.Pattern;
+
+public class FormattingVisitor extends PsiRecursiveElementWalkingVisitor {
+    private static final Pattern nonWordPattern = Pattern.compile("\\W*");
+    StringBuilder sb = new StringBuilder();
+    boolean isPreviousSpaceSensitive = false;
+
+    @Override
+    public void visitElement(@NotNull PsiElement element) {
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/FormattingVisitor.java:34]: add a comment, please

File: refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java (hunk 5 above)
Comment (onewhl) [reply refactoringminer/src/main/java/gr/uom/java/xmi/LocationInfo.java:63]: add todo about removing this method when migration if finished

File: refactoringminer/src/main/java/gr/uom/java/xmi/UMLType.java [hunk 9]
```diff
@@ -1,167 +1,253 @@
...
-                rightType.annotations.add(new UMLAnnotation(cu, filePath, annotation));
+        }
+        if (left == null) {
+            left = extractGenerics(qualifiedName);
+        }
+        return left;
+    }
+
+    @NotNull
+    private static List<String> splitCapturing(String qualifiedName, char separator) {
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/UMLType.java:83]: add a comment, please

File: refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/OperationBody.java [hunk 10]
```diff
@@ -77,220 +81,236 @@ public VariableDeclaration getVariableDeclaration(String variableName) {
...
-            VariableDeclaration vd = new VariableDeclaration(cu, filePath, variableDeclaration);
+            PsiParameter variableDeclaration = foreachStatement.getIterationParameter();
+            VariableDeclaration vd = new VariableDeclaration(file, filePath, variableDeclaration);
             child.addVariableDeclaration(vd);
-            AbstractExpression variableDeclarationName = new AbstractExpression(cu, filePath, variableDeclaration.getName(), CodeElementType.ENHANCED_FOR_STATEMENT_PARAMETER_NAME);
+            AbstractExpression variableDeclarationName = new AbstractExpression(file, filePath, variableDeclaration.getNameIdentifier(), CodeElementType.ENHANCED_FOR_STATEMENT_PARAMETER_NAME);
             child.addExpression(variableDeclarationName);
-            if (variableDeclaration.getInitializer() != null) {
-                AbstractExpression variableDeclarationInitializer = new AbstractExpression(cu, filePath, variableDeclaration.getInitializer(), CodeElementType.VARIABLE_DECLARATION_INITIALIZER);
+            assert variableDeclaration.getInitializer() == null;
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/OperationBody.java]: please, take a look at it

File: refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java [hunk 11]
```diff
@@ -36,707 +42,264 @@
...
+                    if (!(declaredElement.getParent() instanceof PsiLambdaExpression)) {
+                        VariableDeclaration variableDeclaration =
+                            new VariableDeclaration(file, filePath, (PsiVariable) declaredElement);
+                        variableDeclarations.add(variableDeclaration);
+                        onLastAnonymous(anonymous -> anonymous.getVariableDeclarations().add(variableDeclaration));
+                    }
+                } else if (declaredElement instanceof PsiClass) {
+                    // Local classes
+                    goInSubtree = false;
+                    // TODO: why local classes are ignored?
```
Comment (onewhl) [inline refactoringminer/src/main/java/gr/uom/java/xmi/decomposition/Visitor.java]: take a look at it please

**Comment by onewhl** [comment]
Well done, thank you!

**Pull request merged**